### PR TITLE
[SPARK-43930][SQL][PYTHON][CONNECT] Add unix_* functions to Scala and Python

### PR DESCRIPTION
**Pull request by panbingkun**
### What changes were proposed in this pull request?
Add following functions:

- unix_date
- unix_micros
- unix_millis
- unix_seconds

to:

- Scala API
- Python API
- Spark Connect Scala Client
- Spark Connect Python Client

### Why are the changes needed?
for parity

### Does this PR introduce _any_ user-facing change?
Yes, new functions.

### How was this patch tested?
- Add New UT.

### CODE REVIEW

**Comment by panbingkun** [comment]
cc @zhengruifeng

File: sql/core/src/test/scala/org/apache/spark/sql/DateFunctionsSuite.scala [hunk 1]
```diff
@@ -545,6 +545,50 @@ class DateFunctionsSuite extends QueryTest with SharedSparkSession {
...
 
+  test("function unix_date") {
+    val d1 = Date.valueOf("2015-07-22")
+    val d2 = Date.valueOf("2015-07-01")
+    val d3 = Date.valueOf("2014-12-31")
+    val df = Seq(d1, d2, d3).toDF("d")
+
+    checkAnswer(
+      df.select(unix_date(col("d"))),
+      Seq(Row(16435), Row(16617), Row(16638)))
```
Comment (zhengruifeng) [inline sql/core/src/test/scala/org/apache/spark/sql/DateFunctionsSuite.scala:556]: can we compare the result with `df.selectExpr("unix_date(d)")`? the numbers here are hard to understand

File: sql/core/src/test/scala/org/apache/spark/sql/DateFunctionsSuite.scala [hunk 2]
```diff
@@ -545,6 +545,50 @@ class DateFunctionsSuite extends QueryTest with SharedSparkSession {
...
+
+  test("function unix_micros") {
+    val t1 = Timestamp.valueOf("2015-07-22 10:00:00")
+    val t2 = Timestamp.valueOf("2014-12-31 23:59:59")
+    val t3 = Timestamp.valueOf("2014-12-31 23:59:59")
+    val df = Seq(t1, t2, t3).toDF("t")
+
+    checkAnswer(
+      df.select(unix_micros(col("t"))),
+      Seq(Row(1420099199000000L), Row(1420099199000000L), Row(1437584400000000L)))
```
Comment (zhengruifeng) [inline sql/core/src/test/scala/org/apache/spark/sql/DateFunctionsSuite.scala:567]: ditto

File: sql/core/src/test/scala/org/apache/spark/sql/DateFunctionsSuite.scala [hunk 3]
```diff
@@ -545,6 +545,50 @@ class DateFunctionsSuite extends QueryTest with SharedSparkSession {
...
+
+  test("function unix_millis") {
+    val t1 = Timestamp.valueOf("2015-07-22 10:00:00")
+    val t2 = Timestamp.valueOf("2014-12-31 23:59:59")
+    val t3 = Timestamp.valueOf("2014-12-31 23:59:59")
+    val df = Seq(t1, t2, t3).toDF("t")
+
+    checkAnswer(
+      df.select(unix_millis(col("t"))),
+      Seq(Row(1420099199000L), Row(1420099199000L), Row(1437584400000L)))
```
Comment (zhengruifeng) [inline sql/core/src/test/scala/org/apache/spark/sql/DateFunctionsSuite.scala:578]: ditto

File: sql/core/src/test/scala/org/apache/spark/sql/DateFunctionsSuite.scala [hunk 4]
```diff
@@ -545,6 +545,50 @@ class DateFunctionsSuite extends QueryTest with SharedSparkSession {
...
+
+  test("function unix_seconds") {
+    val t1 = Timestamp.valueOf("2015-07-22 10:00:00")
+    val t2 = Timestamp.valueOf("2014-12-31 23:59:59")
+    val t3 = Timestamp.valueOf("2014-12-31 23:59:59")
+    val df = Seq(t1, t2, t3).toDF("t")
+
+    checkAnswer(
+      df.select(unix_seconds(col("t"))),
+      Seq(Row(1420099199L), Row(1420099199L), Row(1437584400L)))
```
Comment (zhengruifeng) [inline sql/core/src/test/scala/org/apache/spark/sql/DateFunctionsSuite.scala:589]: ditto

File: python/pyspark/sql/functions.py [hunk 5]
```diff
@@ -4981,6 +4981,64 @@ def to_date(col: "ColumnOrName", format: Optional[str] = None) -> Column:
...
+@try_remote_functions
+def unix_date(col: "ColumnOrName") -> Column:
+    """Returns the number of days since 1970-01-01.
+
+    Examples
+    --------
+    >>> spark.conf.set("spark.sql.session.timeZone", "America/Los_Angeles")
+    >>> df = spark.createDataFrame([('1970-01-02',)], ['t'])
+    >>> df.select(unix_date(to_date(df.t)).alias('n')).collect()
+    [Row(n=1)]
```
Comment (zhengruifeng) [inline python/pyspark/sql/functions.py:4993]: should unset the config after this doctest, in order to make sure not pollute the env
```
spark.conf.unset("spark.sql.session.timeZone")
```

File: python/pyspark/sql/functions.py [hunk 6]
```diff
@@ -4981,6 +4981,64 @@ def to_date(col: "ColumnOrName", format: Optional[str] = None) -> Column:
...
+@try_remote_functions
+def unix_micros(col: "ColumnOrName") -> Column:
+    """Returns the number of microseconds since 1970-01-01 00:00:00 UTC.
+
+    Examples
+    --------
+    >>> spark.conf.set("spark.sql.session.timeZone", "America/Los_Angeles")
+    >>> df = spark.createDataFrame([('2015-07-22 10:00:00',)], ['t'])
+    >>> df.select(unix_micros(to_timestamp(df.t)).alias('n')).collect()
+    [Row(n=1437584400000000)]
```
Comment (zhengruifeng) [inline python/pyspark/sql/functions.py:5007]: ditto

File: python/pyspark/sql/functions.py [hunk 7]
```diff
@@ -4981,6 +4981,64 @@ def to_date(col: "ColumnOrName", format: Optional[str] = None) -> Column:
...
+def unix_millis(col: "ColumnOrName") -> Column:
+    """Returns the number of milliseconds since 1970-01-01 00:00:00 UTC.
+    Truncates higher levels of precision.
+
+    Examples
+    --------
+    >>> spark.conf.set("spark.sql.session.timeZone", "America/Los_Angeles")
+    >>> df = spark.createDataFrame([('2015-07-22 10:00:00',)], ['t'])
+    >>> df.select(unix_millis(to_timestamp(df.t)).alias('n')).collect()
+    [Row(n=1437584400000)]
```
Comment (zhengruifeng) [inline python/pyspark/sql/functions.py:5022]: ditto

File: python/pyspark/sql/functions.py [hunk 8]
```diff
@@ -4981,6 +4981,64 @@ def to_date(col: "ColumnOrName", format: Optional[str] = None) -> Column:
...
+def unix_seconds(col: "ColumnOrName") -> Column:
+    """Returns the number of seconds since 1970-01-01 00:00:00 UTC.
+    Truncates higher levels of precision.
+
+    Examples
+    --------
+    >>> spark.conf.set("spark.sql.session.timeZone", "America/Los_Angeles")
+    >>> df = spark.createDataFrame([('2015-07-22 10:00:00',)], ['t'])
+    >>> df.select(unix_seconds(to_timestamp(df.t)).alias('n')).collect()
+    [Row(n=1437584400)]
```
Comment (zhengruifeng) [inline python/pyspark/sql/functions.py:5037]: ditto

Comment (panbingkun) [reply python/pyspark/sql/functions.py:5040]: This is all done!

**Review by zhengruifeng — APPROVED** [review]

**Comment by zhengruifeng** [comment]
merged to master

**Pull request closed without merging**